core/dbt/rpc.py

```diff
@@ -66,3 +90,306 @@ def terminating(cls):
+        except dbt.exceptions.RPCKilledException as exc:
+            # do NOT log anything here, you risk triggering a deadlock on the
+            # queue handler we inserted above
+            error = dbt_error(exc)
+        except dbt.exceptions.Exception as exc:
+            logger.debug('dbt runtime exception', exc_info=True)
+            error = dbt_error(exc)
+        except Exception as exc:
+            logger.debug('uncaught python exception', exc_info=True)
+            error = server_error(exc)
+
+        # put whatever result we got onto the queue as well.
+        if error is not None:
+            self.queue.put([QueueMessageType.Error, error.error])
+        else:
+            self.queue.put([QueueMessageType.Result, result])
+
+    def handle(self, kwargs):
+        self.started = time.time()
+        self.timeout = kwargs.pop('timeout', None)
+        self.queue = multiprocessing.Queue()
+        self.process = multiprocessing.Process(
+            target=self.task_bootstrap,
+            args=(kwargs,)
+        )
+        self.process.start()
+        return self.get_result()
+
+    @property
+    def state(self):
+        if self.started is None:
+            return 'not started'
+        elif self.process is None:
+            return 'initializing'
+        elif self.process.is_alive():
+            return 'running'
+        else:
+            return 'finished'
+
+
+TaskRow = namedtuple(
+    'TaskRow',
+    'task_id request_id request_source method state start elapsed timeout'
+)
+
+
+class TaskManager(object):
+    def __init__(self):
+        self.tasks = {}
+        self.completed = {}
+        self._rpc_task_map = {}
+        self._rpc_function_map = {}
+        self._lock = multiprocessing.Lock()
+
+    def add_request(self, request_handler):
+        self.tasks[request_handler.task_id] = request_handler
+
+    def add_task_handler(self, task):
+        self._rpc_task_map[task.METHOD_NAME] = task
+
+    def rpc_task(self, method_name):
+        return self._rpc_task_map[method_name]
+
+    def process_listing(self, active=True, completed=False):
+        included_tasks = {}
+        with self._lock:
+            if completed:
+                included_tasks.update(self.completed)
+            if active:
+                included_tasks.update(self.tasks)
+
+        table = []
+        now = time.time()
+        for task_handler in included_tasks.values():
+            start = task_handler.started
+            if start is not None:
+                elapsed = now - start
+
+            table.append(TaskRow(
+                str(task_handler.task_id), task_handler.request_id,
+                task_handler.request_source, task_handler.method,
+                task_handler.state, start, elapsed, task_handler.timeout
+            ))
+        table.sort(key=lambda r: (r.state, r.start))
+        result = {
+            'columns': list(TaskRow._fields),
+            'rows': [list(r) for r in table],
+        }
```

Comment:
can you make this a list of objects instead? e.g.

```javascript
[{
  "task_id": "1f7054d4-8dae-4560-9cc1-662127de4184",
  "request_id": "5bf9dec2-431b-44e2-9d76-1239928571ad",
  "request_source": "10.1.31.102",
  "method": "run",
  "state":  "running",
  "start": 1554303525.1364384,
  "elapsed": 0.014189481735229492,
  "timeout": 900
}]
```

In the frontend, I have to iterate over the list of tasks and match the request id I sent to the request id in the table. if the indexing of the table changes, my loop will break.... but this way I can just look the task up by key.

if `data` has to be a dict, just put this list under another key, up to you what that is called

Reply:
I'm pretty sure it does have to be a dict, so I named the key `rows`.